blueprints/to-locks.md

```diff
@@ -0,0 +1,279 @@
+| shared      | boolean                     | required    | whether or not this is a shared lock                                                    |
+| creator     | boolean                     | required    | whether or not the requesting `username` is the first one to acquire a lock on `cdn`| 
+| lastUpdated | time                        | required    | the last time this lock was updated                                                     |
+
+**API constraints:**
+- a `username` and `cdn` combination must be unique if `shared` is set to `false`. In other words, there can be only one `exclusive`
+lock for a CDN
+- a CDN can have multiple `shared` locks
+- the `creator` will be set to `true` if the user is the first one acquiring a lock for the specified `cdn`
+- a user can snap/ queue a CDN only if the `creator` field corresponding to the `username` and `cdn` combination is `true`
+- a user can delete their `exclusive` lock whenever, since the user is the creator in this case
+- a user who is the creator cannot delete their `shared` lock until all other non-creator users have releases their shared locks for that CDN
+- a user with at least `all-read` and `all-write` capabilities can delete the locks of other users on any CDN. The `all-write` capability will be modified
+to include a new capability `delete-all-locks`.
+
+Three new endpoints will be added for `GET`, `POST` and `DELETE` functionality with respect to CDN locks.
+##### GET `cdn_locks`
+
+response JSON:
+```JSON
+{
+  "response": [
+    {
+      "username": "foo",
+      "cdn": "cdn1",
+      "message": "snapping cdn",
+      "shared": true,
+      "creator": true,
+      "lastUpdated": "2021-05-10T16:03:34-06:00"
+    },
+    {
+      "username": "bar",
+      "cdn": "cdn2",
+      "message": "queue cdn",
+      "shared": false,
+      "creator": true,
+      "lastUpdated": "2021-05-10T17:04:34-06:00"
+    }
+  ]
+}
+```
+
+##### POST `cdn_locks`
+
+request JSON:
+```JSON
+{
+  "cdn": "bar",
+  "message": "snapping cdn",
+  "shared": false
+}
+```
+
+response JSON:
+```JSON
+{
+  "alerts": [
+    {
+      "text": "CDN lock acquired!",
+      "level": "success"
+    }
+  ],
+  "response": {
+    "username": "foo",
+    "cdn": "bar",
+    "message": "snapping cdn",
+    "shared": false,
+    "creator": true,
+    "lastUpdated": "2021-05-10T17:05:30-06:00"
+  }
+}
+```
+
+##### `DELETE /cdn_locks?cdn=bar`
+
+response JSON:
+```JSON
+{
+  "alerts": [
+    {
+      "text": "Cdn lock deleted",
+      "level": "success"
+    }
+  ],
+  "response": {
+    "username": "foo",
+    "cdn": "bar",
+    "message": "snapping cdn",
+    "shared": false,
+    "creator": true,
+    "lastUpdated": "2021-05-10T17:05:30-06:00"
+  }
+}
+```
+
+#### Client Impact
+
+New Go client methods will be added for the `/cdn_locks` endpoints in order to write TO API tests for the new endpoints.
+
+#### Data Model / Database Impact
```

Comment:
I didn't see any discussion of the actual implementation, so I'm curious. 

How will the lock be implemented? Will you be using Postgres Advisory locks? If not, are there particular atomic operations you'll make use of to avoid race conditions in acquiring, evaluating, and releasing locks?

Reply:
So the lock object is just another database table. So when a snap/ queue request comes in, we just do a look up in the `cdn_lock` table to see if the current user has the lock, and if so, proceed with the operation. In the current scenarion, this is just like checking for the `If-Modified-Since` or `If-Unmodified-Since` header.

Reply:
What prevents two users from obtaining a lock on the same resource?

Reply:
When a user sends a request to snap/ queue a CDN, the backend code for API endpoint does a database check to see if the current user has a lock on the CDN. If not, then a `403 Forbidden` is returned to the user.
The API uniqueness constraint (of `cdn` in case of an exclusive lock and `username`+`cdn` in case of a shared lock) will prevent two users obtaining the lock on the same resource. We also suggest the clients make use of the `If-Unmodified-Since` header to make sure the locks data doesn't change on them before they complete their request.

Reply:
Is either of these examples possible?

```
User1 API request received to acquire exclusive lock. API checks database, no existing lock
User2 API request received to acquire exclusive lock. API checks database, no existing lock
User1 writes lock to DB. Returns lock via API
User2 writes lock to DB. Returns lock via API
```

```
(User1, cdn) lock in db
User1 sends API request to snap. API checks database, lock exists
Before snap completes, user1 releases lock, removing entry from db table
User2 modifies CDN
User1 snap resumes and includes User 2 modifications (bad?)
```

Reply:
In the first scenario, if user 2 uses `If-UnModified-Since` in their request (which we can make a requirement in the backend).
For the second scenario, we can add a mutex or some kind of sync mechanism to make sure that a lock cannot be released by the "owner" when a snap is in progress. Frankly, I think thats a very rare edge case, because the snap usually takes a very small amount of time, and you would really have to try hard to release the lock between the time you press the `snap` button and the snap completes.
Note that when `user1` has a lock on `cdn` (assuming `user1` is the owner of the lock), only `user1` can snap/ queue, so it won't really matter if another user that has the shared lock on `cdn` releases the lock during this time or not. Hope that makes sense.